tools/topology/topology2/platform/intel/sdw-amp-generic.conf

```diff
@@ -54,6 +55,19 @@ IncludeByKey.PASSTHROUGH {
 				Object.Widget.alh-copier.1 {
 					stream_name $SDW_SPK_STREAM
 					node_type $ALH_LINK_OUTPUT_CLASS
+					IncludeByKey.SDW_AMP_FMT_24 {
```

Comment:
@bardliao I'm a bit puzzled by this. The kernel has special logic if only one format is defined by copier (DAI link format mask is forced to this format in this case). But why doesn't this work with multiple formats in copier. If the codec only supports S24_LE (32bit-with-24validbits), format should be constrained already.

@ranj063 Is this correct? Looking at kernel sof_ipc4_prepare_copier_module(), dai copier should always be limited to single output format (as there's no logic to decide which one to use if there is multiple).

Reply:
> If the codec only supports S24_LE (32bit-with-24validbits), format should be constrained already.

@kai In theory, yes. But, somehow AlSA may still select 32 bit format even the codec driver declares it does not support 32 bit format. That's why we call `snd_mask_set_format` in `sof_ipc4_pcm_dai_link_fixup` to specify the BE format.

Reply:
This change fixes output format to S24_LE. why ?  Kernel driver can't choose S24_LE in 3 format array ?

Reply:
> This change fixes output format to S24_LE. why ? Kernel driver can't choose S24_LE in 3 format array ?

Because cs42l43 and cs35l56 can't support 32 bit format. So that we want to use 24 bit format.

Reply:
Chatted with @RanderWang, alternative way is to use 24 bit format for all sdw codecs. There seems to be no much benefit to use 32 bit format. What do you think? @kv2019i @ranj063 @plbossart

Reply:
doesn't the Intel hardware require 32 bits anyways, so why would we restrict the number of valid bits?

Reply:
> doesn't the Intel hardware require 32 bits anyways, so why would we restrict the number of valid bits?

@plbossart The purpose is to let kernel to use the 24 bit format. The related kernel change is https://github.com/thesofproject/linux/pull/4446

Reply:
@kv2019i I need to clarify that the format here means the BE format. Based on my experiment, ALSA will use the audio file's format if no one try to set the format. That is the case for IPC4 before https://github.com/thesofproject/linux/pull/4446 was merged. For example, it will be 32 bit format if we play a 32 bit format .wav file, and it will be 24 bit format if we play a 24 bit .wav file.
I agree that ALSA should not use the format that it is not supported by the codec. Maybe we can add a mechanism to select the format that is supported by both cpu and codec. But it will be a big task.
We set the BE format if there is only one format available in the dai copier after https://github.com/thesofproject/linux/pull/4446 was merged. 
And we have set the BE format in IPC3 in  [sof_ipc3_pcm_dai_link_fixup](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc3-pcm.c#L237) for a while.

Reply:
@kv2019i Yes, always using 24 bit BE format makes sense, and much simpler. The only open in my mind is that how does kernel know a specific codec want to use 32 bit format?
I looked at `hdac_hda_dai_hw_params`, and found it is not my intention. The params_format(params) is still S32_LE, not S24_LE although `maxbps` is 24. It is more like that ASoC says hey let's use S32_LE format, but the HDA codec driver changes the format by itself. I don't think we can do it for all SDW/I2S codec drivers.
IMHO, the right way is let params_format(params) become S24_LE, and codec drivers will set the format configuration according to params_format(params).
After removing `snd_mask_none` and `snd_mask_set_format` from `sof_ipc3_pcm_dai_link_fixup`, the params_format(params) became the same format as what I play. I need some time to figure out where does params_format(params) be decided, but I think using `snd_mask_set_format` is the right way to config the BE format.
Back to this PR, what would you prefer?
1. Change the default SDW format to S24_LE in topology, and kernel set the BE format according to topology's format.
2. Kernel always set S24_LE format for SDW. Topology can use any format it wants and the format will be used in SOF.